pandas/core/frame.py

```diff
@@ -3891,6 +3892,22 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
         if not isinstance(keys, list):
             keys = [keys]
 
+        missing = []
+        for x in keys:
+            if not (is_scalar(x) or isinstance(x, tuple)):
+                if not isinstance(x, (ABCSeries, ABCIndexClass, ABCMultiIndex,
```

Comment:
you can use is_list_like here

Reply:
This would also change the capabilities of `set_index`. e.g. an iterator has `is_list_like = True` - do you want to support that? Currently, only the list that I'm instance-checking (plus column keys) are allowed.

Reply:
I don't want to explicity list types here as its verbose and fragile, use is_list_like and not is_iterator if you must (though to be honest an iterator is actually ok)

Reply:
> I don't want to explicity list types here as its verbose and fragile

I have a feeling the implementation you request will be much more verbose, but OK

> OK. How to deal with tuples then. Can be valid column keys and valid list-likes.

You want to try tuples as column keys first and *then* as an array (would make the most sense to me), or something else?